### PR TITLE
introduce sq8 uniform quantizer

### PR DESCRIPTION
**Pull request by LHT129**

### CODE REVIEW

**Review by inabao — APPROVED** [review]
LGTM

File: src/quantization/sq8_uniform_quantizer.h [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+    uint64_t offset_sum_{0};
+};
+
+template <MetricType metric>
+SQ8UniformQuantizer<metric>::SQ8UniformQuantizer(int dim, Allocator* allocator)
+    : Quantizer<SQ8UniformQuantizer<metric>>(dim, allocator) {
+    lower_bound_ = std::numeric_limits<DataType>::max();
+    diff_ = std::numeric_limits<DataType>::lowest();
+
+    this->code_size_ = 0;
```
Comment (jiaweizone) [inline src/quantization/sq8_uniform_quantizer.h:92]: minor: `this` is not necessary?

File: src/quantization/sq8_uniform_quantizer.h [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+        sum_type sum2 = *((sum_type*)(codes2 + offset_sum_));
+
+        result = lower_bound_ * (sum1 + sum2) + (diff_ / 255.0) * (diff_ / 255.0) * result -
+                 lower_bound_ * lower_bound_;
+
+        result = 1 - result;
+
+    } else {
+        logger::error("unsupported metric type");
+        result = 0;
```
Comment (jiaweizone) [inline src/quantization/sq8_uniform_quantizer.h:224]: `0` can indicate an error ?

File: src/quantization/sq8_uniform_quantizer.h (hunk 1 above)
Comment (LHT129) [reply src/quantization/sq8_uniform_quantizer.h:92]: for template class this is necessary for var in basic class

File: src/quantization/sq8_uniform_quantizer.h (hunk 2 above)
Comment (LHT129) [reply src/quantization/sq8_uniform_quantizer.h:224]: same behavior as sq4 uniform, actually should process the wrong metric by throw exception

File: src/quantization/sq8_uniform_quantizer.h (hunk 1 above)
Comment (jiaweizone) [reply src/quantization/sq8_uniform_quantizer.h:92]: OK，for template class is necessary

File: src/quantization/sq8_uniform_quantizer.h (hunk 2 above)
Comment (jiaweizone) [reply src/quantization/sq8_uniform_quantizer.h:224]: Do we need a special state to indicate failure ？

Comment (LHT129) [reply src/quantization/sq8_uniform_quantizer.h:224]: actually the metric error have special state at param check. in this code, we will not meet the metric error, but as an individual component, we suggest quantizer can process all errors by itself. So in future, we will have specific state and method to deal with this errors

Comment (jiaweizone) [reply src/quantization/sq8_uniform_quantizer.h:224]: OK

**Review by jiaweizone — APPROVED** [review]
LGTM

**Review by ShawnShawnYou — APPROVED** [review]

**Pull request merged**